src/logger.js

```diff
@@ -45,4 +69,105 @@ export default once((config) => {
     });
 
     return logger;
-});
+}
+
+// filter out named properties from req object
+function omit(req, blacklist) {
+    return omitBy(req, (value, key) => blacklist.includes(key));
+}
+
+// exclude any health or other ignorable urls
+function skip(config) {
+    return function ignoreUrl(req) {
+        const url = req.originalUrl || req.url;
+        return config.ignoreRouteUrls.includes(url);
+    };
+}
+
+// where morgan connects to winston
+function addStream(logger, config) {
+    return {
+        write: (message) => {
+            const logEntry = JSON.parse(message);
+            return logger.log(config.level, logEntry.message, logEntry);
+        },
+    };
+}
+
+
+@binding('logger')
+@defaults(loggingDefaults)
+class Logger {
+    constructor(graph) {
+        const config = graph.config.logger;
+        const baseLogger = createLogger(config);
+        this.baseLogger = baseLogger;
+        this.requestRules = config.requestRules;
+
+        // define custom tokens
+        morgan.token('operation-hash', req => get(req, 'body.extensions.persistentQuery.sha256Hash'));
+        morgan.token('operation-name', req => get(req, 'body.operationName'));
+        morgan.token('user-id', req => get(req, 'user.id'));
+        morgan.token('message', req => req.name || '-');
+        morgan.token('request-id', req => req.id);
+        morgan.token('request-headers', (req) => {
+            const headers = config.includeReqHeaders === true
+                ? omit(req.headers, config.omitReqProperties) : {};
+            return JSON.stringify(headers);
+        });
+
+        const format = json(config.morgan.format);
+        const options = {
+            stream: addStream(this.baseLogger, config),
+            skip: skip(config),
+        };
+        graph.app.use(morgan(format, options));
```

Comment:
I don't love that I attached the `app` to the graph in my main `index.js` but it lets you add midlewares easily enough....